docs/man/scxmlrun.man
'\" t
.\" Manual page created with latex2man on Wed Mar 27 11:39:54 2019
.\" NOTE: This file is generated, DO NOT EDIT.
.de Vb
.ft CW
.nf
..
.de Ve
.ft R

.fi
..
.TH "SCXMLRUN" "1" "November 2018" "LDL Tools " "LDL Tools "
.SH NAME

scxmlrun \-\- MQTT\-enabled SCXML interpreter 
.PP
.SH SYNOPSIS

scxmlrun <option>* <scxmlfile>? <infile>? 
.PP
.SH DESCRIPTION

scxmlrun reads \fI<scxmlfile>\fP,
and then processes 
a sequence of input events listed in \fI<infile>\fP
as instructed in \fI<scxmlfile>\fP\&.
Instead of reading from \fI<infile>\fP,
input events can be received over the network via MQTT. 
Output events emitted using \fISCXML.send\fP
in \fI<scxmlfile>\fP
can be 
either written to a local file or transmitted via MQTT. 
.PP
.SH OPTIONS

.TP
\fB\-\-model\fP <scxmlfile>
 reads a state\-chart in SCXML from \fI<scxmlfile>\fP
.PP
.TP
\fB\-i\fP <infile>
 read (lines of) SCXML events (in JSON) from \fI<infile>\fP
.TP
\fB\-o\fP <outfile>
 write output events (in JSON) to \fI<outfile>\fP
.TP
\fB\-\-trace\fP <tracefile>
 write trace info to \fI<tracefile>\fP\&.
Trace reports which states execution visits through. 
.PP
.TP
\fB\-\-mqtt[:in|:out]\fP
 assume MQTT\-based I/O by default.
.br
\fB\-\-mqtt:in\fP,
\fB\-\-mqtt:out\fP,
or \fB\-\-mqtt\fP
assumes MQTT for input, output, or both, respectively. 
.PP
.TP
\fB\-\-broker\fP <host>[:<port>]
 specifiy \fI<host>\fP
(at \fI<port>\fP)
as the MQTT broker to connect to.
.br
(default: localhost:1883) 
.TP
\fB\-\-sub\fP <topic>
 subscribe to \fI<topic>\fP
for input events.
.br
multiple topics can be specified as: \-\-sub t1 \-\-sub t2 \&.\&.\&. 
.TP
\fB\-\-pub\fP <topic>
 specify \fI<topic>\fP
as the default topic for output events. 
This is used by \fISCXML.send\fP
in \fI<scxmlfile>\fP
when no topic is passed as a parameter. 
.TP
\fB\-\-trace\-pub\fP <topic>
 specify \fI<topic>\fP
for publishing trace info 
.PP
.TP
\fB\-r\fP, \fB\-\-relay\fP
 run \fBscxmlrun\fP
as a MQTT \fIrepeater\fP,
that is, 
\fBscxmlrun\fP
reads input events and transmits them without any interpretation. 
No SCXML model should be specified in this mode. 
.PP
.TP
\fB\-v\fP, \fB\-\-verbose\fP
 become verbose. 
.TP
\fB\-q\fP, \fB\-\-silent\fP
 stay quiet. 
.TP
\fB\-V\fP, \fB\-\-version\fP
 show version info. 
.TP
\fB\-h\fP, \fB\-\-help\fP
 show usage. 
.PP
.SH EXAMPLES

Consider the following simple SCXML model, which we here call "echo.scxml". 
.PP
.Vb
<scxml xmlns="http://www.w3.org/2005/07/scxml" version="1.0" datamodel="ecmascript" initial="q1">
  <state id="q1">
    <transition target="q2" event="echo">
      <script>console.log (_event.data)</script>
    </transition>
  </state>
  <final id="q2"/>
</scxml>
.Ve
.PP
.TP
1.
reading event(s) from a local file 
.PP
$ echo \&'{"event":{"name":"echo","data":"hello"}}\&' | scxmlrun echo.scxml
.br
hello
.PP
.TP
2.
reading event(s) via MQTT 
.PP
$ scxmlrun echo.scxml \-\-sub echo
.br
$ mosquitto_pub \-t echo \-m \&'{"event":{"name":"echo","data":"hello"}}\&'
.br
hello
.PP
.SH INPUT/OUTPUT EVENT FORMAT

.PP
Input and output of \fBscxmlrun\fP
are SCXML events in the JSON format.
.br
Each event is a JSON object of the form: 
.ce 100
{"event":{"name":...,\&.\&.\&.},"type":"mqtt", "topic":..., \&.\&.\&.}
.ce 0

.PP
See \fBhttps://www.w3.org/TR/scxml/#send\fP
for the detail. 
.PP
Note that, as a \fBscxmlrun\fP\-specific
feature, 
"topic"\-key and its value can be included in each input/output event, 
which is referred to when connecting to the MQTT broker. 
.PP
.SH ENVIRONMENT VARIABLES

.PP
\fIMQTT_HOST\fP
is looked up, when no broker is explicitly specified. 
.PP
.SH SEE ALSO

rules2scxml 
.PP
.SH AUTHOR

LDLTools development team at IBM Research. 
.PP
.TP
.B *
URL: \fBhttps://ldltools.github.io\fP
.TP
.B *
Email: \fBldltools@outlook.com\fP
.PP
.SH COPYRIGHT

(C) Copyright IBM Corp. 2018. 
License Apache 2.0.
.br
.PP
This is free software: you are free to change and redistribute it. 
There is NO WARRANTY, to the extent permitted by law. 
.PP
.\" NOTE: This file is generated, DO NOT EDIT.


CHANGES.md
# v0.12.0 (2019-03.29)

- **new** [scxmlrun-all](tools/runall):
  run several SCXML processes (along with generic MQTT processes) in parallel.

- add `*_mqtt.conf` for testing composite SCXML processes connected via MQTT.
- update/adjust examples

# v0.11.0 (2018-12-27)

- **new** introduce _"event repeater"_ (activated by: `scxmlrun -r`) which relays incoming events to elsewhere.

- add/update examples
- add `*.conf` files for unit-testing using [shelltest](https://github.com/simonmichael/shelltestrunner)
- add man page (for `man scxmlrun`)

# v0.10.0 (2018-11-08)

- **new** introduce JS functions: `SCXML.raise` and `SCXML.send`

- add examples: microwave, calc (with web ui), traffic-control, and blackjack
- add examples: voting and vulnerable

# v0.9.0 (2018-10-06)

initial public release


src/scxmlrun.cpp
// (C) Copyright IBM Corp. 2018.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "scxmlinterpreter.hpp"
#include "mqttrepeater.hpp"
#include "version.hpp"

#ifdef USE_QTSCXML
#include "qtscxml/qtscxmlproc.hpp"
#include <QtCore/QCoreApplication>
#define SCXMLPROC scxml::qtscxmlproc
#endif

#ifdef USE_USCXML
#include "uscxml/uscxmlproc.hpp"
#define SCXMLPROC scxml::uscxmlproc
#endif

#include <cstdio>
#include <codecvt>
#include <iostream>
#include <list>
#include <sstream>
#include <string>
#include <cstdlib>
#include <cassert>
//#include <thread>

static void
synopsis (const char* prog)
{
    std::cout << prog << " v" << SCXMLRUN_VERSION << "\n";
    std::cout << "usage: " << prog << " <option>* <scxmlfile>? <infile>?\n";
    std::cout << "options\n";
    std::cout << "  -m <scxmlfile>\t"
              << "read scxml from <scxmlfile>\n";
    std::cout << std::endl;

    std::cout << "  -i <infile>\t\t"
              << "read input events from <infile>\n";
    std::cout << "  -o <outfile>\t\t"
              << "write output events to <outfile>\n";
    std::cout << "  --trace <tracefile>\t"
              << "write trace to <tracefile>\n";
    std::cout << std::endl;

    std::cout << "  --mqtt\t\t"
              << "assume MQTT-based I/O by default\n";
    std::cout << "  --broker <host>\t"
              << "connect to <host> as the MQTT broker\n";
    std::cout << "  --sub <topic>\t\t"
              << "subscribe to <topic> for input events\n";
    std::cout << "  --pub <topic>\t\t"
              << "publish output events to <topic> by default\n";
    std::cout << "  --trace-pub <topic>\t"
              << "publish trace to <topic>\n";
    std::cout << std::endl;

    std::cout << "  -h, --help\t\t"
              << "display this message\n";
    std::cout << "  -v, --verbose\t\t"
              << "become verbose\n";
    std::cout << "  -q, --silent\t\t"
              << "stay quiet\n";
    std::cout << "  -V, --version\t\t"
              << "show version info\n";
}

static void
extra_synopsis (void)
{
    std::cout << std::endl;
    std::cout << "  [advanced/experimental options]" << std::endl;
    std::cout << "  -r, --relay\t\t"
              << "run as a MQTT event repeater\n";
}

static void
mosq_connect (mosquitto*& mosq, const char* host, int port, scxml::interpreter* proc, bool& connected)
{
    if (mosq && connected) return;

    assert (!mosq);
    mosquitto_lib_init ();
    mosq = mosquitto_new (nullptr, true, nullptr);
    proc->mosq_set_callbacks (mosq);
    assert (mosq && !connected);

    //int rslt = mosquitto_connect (mosq, host, 1883, 60);
    int rslt = mosquitto_connect_async (mosq, host, port, 60);
    assert (rslt == MOSQ_ERR_SUCCESS);
    connected = true;
}

static void
mosq_start (mosquitto* mosq, bool& started)
{
    assert (mosq);
    if (started) return;
    
    int rslt = mosquitto_loop_start (mosq);  // threaded
    assert (rslt == MOSQ_ERR_SUCCESS);
    started = true;
}

static void
mosq_terminate (mosquitto* mosq)
{
    if (!mosq) return;

    mosquitto_disconnect (mosq);
    mosquitto_loop_stop (mosq, false);
    mosquitto_destroy (mosq);
    mosquitto_lib_cleanup ();
}

int
main (int argc, char** argv)
{
    const char* scxmlfile = NULL;

    enum {_INTERPRETER, _REPEATER} mode = _INTERPRETER;
      // [remark]
      // in the "repeater" mode, no scxml file is needed/interpreted,
      // but just incoming events are forwarded to elsewhere
    enum {_NONE = 0, _FILE, _MQTT} intype = _NONE, outtype = _NONE, tracetype = _NONE;
    const char* infile = NULL;
    const char* outfile = NULL;
    const char* tracefile = NULL;

    // in/out via MQTT
    const char* mqtt_host = nullptr;
    int mqtt_port = 0;
    std::list<const char*> subs;
    const char* pub = NULL;
    const char* trace_pub = NULL;

    int8_t verbosity = 0;
      // -1: silent
      //  0: default
      //  1: verbose
      //  2: extra-verbose

    for (int i = 1; i < argc; i++)
    {
        if (!strcmp (argv[i], "-m") || !strncmp (argv[i], "--model", 5))
            scxmlfile = argv[++i];

        else if (!strcmp (argv[i], "-i"))
            infile = argv[++i];
        else if (!strcmp (argv[i], "-o"))
            outfile = argv[++i];
        else if (!strcmp (argv[i], "--trace"))
            tracefile = argv[++i];

        else if (!strcmp (argv[i], "--mqtt:in"))
            intype = _MQTT; // this may be changed later
        else if (!strcmp (argv[i], "--mqtt:out"))
            outtype = _MQTT; // this may be changed later
        else if (!strcmp (argv[i], "--mqtt"))
            intype = outtype = tracetype = _MQTT; // this may be changed later
        else if (!strcmp (argv[i], "-b") || !strncmp (argv[i], "--broker", 5))
            mqtt_host = argv[++i];
        else if (!strncmp (argv[i], "--sub", 5))
            subs.push_back (argv[++i]);
        else if (!strncmp (argv[i], "--pub", 5))
            pub = argv[++i];

        else if (!strcmp (argv[i], "-r") || !strncmp (argv[i], "--relay", 5))
        {
            mode = _REPEATER;
            scxmlfile = "/dev/null";
        }

        else if (!strcmp (argv[i], "-v") || !strcmp (argv[i], "--verbose"))
            verbosity = 1;
        else if (!strcmp (argv[i], "-vv"))
            verbosity = 2;
        else if (!strcmp (argv[i], "-q") || !strcmp (argv[i], "--silent"))
            verbosity = -1;
        else if (!strcmp (argv[i], "-V") || !strcmp (argv[i], "--version"))
        {
            SCXMLPROC::version ();
            return 0;
        }
        else if (!strcmp (argv[i], "-h") || !strcmp (argv[i], "--help"))
        {
            synopsis (argv[0]);
            return 0;
        }
        else if (!strcmp (argv[i], "-hh"))
        {
            synopsis (argv[0]); extra_synopsis ();
            return 0;
        }

        else if (*argv[i] == '-')
        {
            std::cerr << "** unknown option: " << argv[i] << std::endl;
            return (1);
        }
        else
            if (!scxmlfile)
                scxmlfile = argv[i];
            else
                infile = argv[i];
    }

    if (!scxmlfile) { synopsis (argv[0]); return (0); }

    // INPUT
    if (infile && subs.size () > 0)
    {
        // more than one input sources: file and mqtt
        std::cerr << "** invalid input specification\n";
        return (-1);
    }
    if (!infile && subs.size () == 0)
        // fall-back
        if (intype != _MQTT)
            infile = "/dev/stdin";
        else
        {
            //subs.push_back ("scxmlrun");
            std::cerr << "** no topic is subscribed to. use \"--sub <topic>\"\n";
            return (-1);
        }
    assert (infile || subs.size () > 0);
    if (infile)
        intype = _FILE;
    else
    {
        assert (subs.size () > 0);
        assert (intype != _FILE);
        intype = _MQTT;
    }

    assert (intype != _NONE);

    // OUTPUT
    if (outfile && pub)
    {
        // more than one output targets: file and mqtt
        std::cerr << "** invalid output specification\n";
        return (-1);
    }
    if (!outfile && !pub)
        // fall-back
        if (outtype != _MQTT)
            outfile = (verbosity < 0) ? "/dev/null" : "/dev/stdout";
        else
            pub = "scxmlrun";
    if (outfile)
        outtype = _FILE;
    if (pub)
        outtype = _MQTT;

    assert (outtype != _NONE);

    // TRACE
    if (tracefile && trace_pub)
    {
        std::cerr << "** invalid trace specification\n";
        return (-1);
    }
    if (!tracefile && !trace_pub)
        // fall-back
        if (tracetype != _MQTT)
            tracefile = (verbosity <= 0) ? "/dev/null" : "/dev/stderr";
        else
            trace_pub = "scxmlrun/trace";
    if (tracefile)
        tracetype = _FILE;
    if (trace_pub)
    {
        tracetype = _MQTT;
    }

    assert (tracetype != _NONE);

    // instantiate proc and load scxml
    scxml::interpreter* proc = nullptr;
    if (mode == _INTERPRETER)
    {
#ifdef USE_QTSCXML
        QCoreApplication* app = new QCoreApplication (argc, argv);
        // non-GUI Qt application
        proc = new SCXMLPROC (app);
#elif defined (USE_USCXML)
        proc = new SCXMLPROC ();
#else
#error "scxml processor is not specified"
#endif
    }
    else if (mode == _REPEATER)
    {
        assert (scxmlfile == "/dev/null");
        proc = new scxml::mqttrepeater ();
    }
    assert (proc);

    // MQTT broker
    mosquitto* mosq = nullptr;
    bool mosq_connected = false, mosq_started =false;

    if (intype == _MQTT || outtype == _MQTT || tracetype == _MQTT)
    {
        if (!mqtt_host)
        {
            mqtt_host = secure_getenv ("MQTT_HOST");  // broker
            if (!mqtt_host) mqtt_host = "localhost";
        }
        assert (mqtt_host);

        const char* found = strchr (mqtt_host, ':');
        // case: <host>:<port>
        if (found)
        {
            sscanf (found, ":%d", &mqtt_port);
            //*found = '\0';
            char* host = (char*) malloc (found - mqtt_host + 1);
            strncpy (host, mqtt_host, found - mqtt_host);
            host[found - mqtt_host] = '\0';
            mqtt_host = host;
        }

        if (mqtt_port == 0)
        {
            const char* mqtt_port_str = secure_getenv ("MQTT_PORT");
            if (mqtt_port_str)
                sscanf (mqtt_port_str, "%d", &mqtt_port);
            else
                mqtt_port = 1883;
        }
        assert (1000 < mqtt_port && mqtt_port < 9999);
    }

    // verbosity
    switch (verbosity)
    {
    case -1:
        std::cout.setstate (std::ios_base::badbit);
        std::cerr.setstate (std::ios_base::badbit);
    case 0:
        std::clog.setstate (std::ios_base::badbit);
        break;
    default:
        {}
    }
    proc->verbosity_set (verbosity);

    // INPUT
    assert (intype != _NONE);
    switch (intype)
    {
    case _FILE:
        proc->eventin_open (infile);
        break;
    case _MQTT:
        {
            mosq_connect (mosq, mqtt_host, mqtt_port, proc, mosq_connected);
            proc->eventin_open (mosq, subs);
            mosq_start (mosq, mosq_started);
        }
        break;
    default:
        assert (false);
        std::cerr << "** input type unspecified\n";
        return (-1);
    }

    // OUTPUT
    assert (outtype != _NONE);
    switch (outtype)
    {
    case _FILE:
        proc->eventout_open (outfile);
        break;
    case _MQTT:
        {
            mosq_connect (mosq, mqtt_host, mqtt_port, proc, mosq_connected);
            proc->eventout_open (mosq, pub);
            mosq_start (mosq, mosq_started);
        }
        break;
    default:
        assert (false);
        std::cerr << "** output type unspecified\n";
        return (-1);
    }

    // TRACE
    switch (tracetype)
    {
    case _FILE:
        proc->traceout_open (tracefile);
        break;
    case _MQTT:
        {
            mosq_connect (mosq, mqtt_host, mqtt_port, proc, mosq_connected);
            proc->traceout_open (mosq, trace_pub);
            mosq_start (mosq, mosq_started);
        }
        break;
    default:
        assert (false);
        std::cerr << "** trace type unspecified\n";
        return (-1);
    }

    // load scxml
    assert (scxmlfile);
    proc->load (scxmlfile);

    // init the engine, etc
    proc->setup ();

    // start processing
    int rslt = 0;
    try
    {
        rslt = proc->run ();
    }
    catch (const std::exception& e)
    {
        std::cerr << "** exception: " << e.what () << std::endl;
        rslt = -1;
    }
    delete proc;
    mosq_terminate (mosq);

    return (rslt);

}
